### ISSUE TITLE
Input formats

### ISSUE THREAD

**Issue by joeroe** [joeroe/stratigraphr#2]
We should be able to handle input data in at least these formats:

- [x] A simple table of relations (i.e. "above", "below", "equal", as on a typical context sheet)
- [ ] Complex tables like @tsdye / `hm`: http://www.tsdye.online/harris-matrix/data-tables/
- [ ] Complex tables like `ArkMatrix`: https://gitlab.com/arklab/ArkMatrix/blob/master/docs/format_csv.md
- [ ] `BASP`/`ArchEd`/`Stratify` .LST files: http://archaeologic.al/wiki/Harris_Matrix#LST

**Comment by nevrome** [comment]
How would you like to implement these data types? A `read.*` function for each type? Or S3 classes that reproduce the respective table structure? Only input or also output?

**Comment by joeroe** [comment]
Good question! I don't think having S3 classes for each would be worthwhile, since they all represent the same thing. What we need are `read_` functions that transform each to a standardised, tabular data structure that describes the stratigraphic relationship between units. That can then be turned into a matrix (an S3 `stratgraph` object) using rules specified by a function like `harris()`.

The question is what the intermediate tabular format should look like. My preference would be something that follows tidy principles, i.e. a single table, where 1 row corresponds to 1 stratigraphic unit and all other observations are columns. Multiple relations can be expressed as nested vectors like those used in `vignette(harris)`.

And yes, now you mention it we should provide `write_` functions for all these too.

**Comment by tsdye** [comment]
Harris' law of stratigraphical succession is satisfied by transitive reduction of the DAG.  The intermediate tabular format might want to serve up information to the transitive reduction function with as little fuss as possible.  Transitive reduction is another feature igraph apparently lacks: [https://github.com/igraph/igraph/issues/932](url).

**Comment by nevrome** [comment]
Ja - the main datatype (and maybe a list class for many of these objects?) has to be defined with all necessary class methods (`is.`, `as.`, `format.`, `print.`). We probably also have to define the accessors `[`, `[<-`, ... to make sure that adding and removing stratigraphic units is handled correctly. 

I like your harris12 sample datatset, @joeroe. List columns are fashionable and a cool feature of R data.frames. If we inherit from tibbles, we already have an excellent base class. 

@tsdye Are you aware of a way to access lisp code from R? I guess it would be great to access the functionality of `hm` for stratigraphr. Could you compile the package to a command line tool? Something that simply runs on any shell after compilation? I just googled for a while without clear results. Lisp is an interpreter language and one always need an interpreter to run it, ja? Just like R?

**Comment by tsdye** [comment]
@nevrome, yes, lisp can produce stand-alone binaries with a command line interface.  It is a bit tricky to write code that works cross-platform, so I'm not sure about producing command line tools for every OS with the `hm` code.  Nevertheless, that is one of my goals for the `hm` project.
I doubt it is possible to call lisp code from R, but I don't know.

**Comment by nevrome** [comment]
@tsdye A command line interface would be great! It's simple enough to call a command line tool from within R. Maybe we could even provide `hm` with the R package as we did with oxcal in [oxcAAR](https://github.com/ISAAKiel/oxcAAR/blob/e84a52f27e7058d6be5d73a1adc67a01c1d6c896/R/utility_functions.R#L51).  

`stratigraphr` could be a simple interface between tidy data and network visualisation in R on the one hand, and Harris graph algorithm implementation in lisp via `hm` on the other. No need to reinvent the wheel. What do you think @joeroe? Is this a way to go?

**Comment by joeroe** [comment]
(The comment I just deleted was a duplicate.)

**Comment by joeroe** [comment]
I'm actually in favour of reinventing wheels, as long as they're not too complicated. It's often easier than having to deal with external dependencies in the long run. But it might be a good stopgap.

**Comment by tsdye** [comment]
Either way is fine with me.  I plan to work on `hm` again this summer, which might or might not suit your plans.

**Comment by nevrome** [comment]
Alright. I could offer to write some reader and writer functions for `stratigraphr` when the main class is defined.

**Comment by joeroe** [comment]
b1daaf32317753a3e7a967a6bb508e5b37d35db4 added a `read_lst()` function, but I only had one file to test it with so it might not be the most robust.

**Comment by joeroe** [comment]
As expected it is not very robust: #9
